runtime/gc_base/modronapi.cpp

```diff
@@ -1089,6 +1089,30 @@ j9gc_notifyGCOfClassReplacement(J9VMThread *vmThread, J9Class *oldClass, J9Class
 	}
 }
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+void
+j9gc_criu_checkPoint(J9VMThread *vmThread)
+{
+	MM_EnvironmentBase *env = MM_EnvironmentBase::getEnvironment(vmThread->omrVMThread);
+	MM_GCExtensionsBase *extensions = env->getExtensions();
+
+	/* trigger a GC to disclaim memory */
+	j9gc_modron_global_collect_with_overrides(vmThread, J9MMCONSTANT_EXPLICIT_GC_SYSTEM_GC);
+	j9gc_modron_global_collect_with_overrides(vmThread, J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT);
+
+	extensions->dispatcher->contractThreadPool(env, extensions->checkpointGCthreadCount);
+}
+
+BOOLEAN
+j9gc_criu_restore(J9VMThread *vmThread)
+{
+	MM_EnvironmentBase *env = MM_EnvironmentBase::getEnvironment(vmThread->omrVMThread);
+	MM_GCExtensionsBase *extensions = env->getExtensions();
+
+	return extensions->dispatcher->expandThreadPool(env, extensions->restoreGCthreadCount);
```

Comment:
I'd generalize this and call a higher level class (like Configuration) and method like reinitializeOnRestore.

That method in turn will call all necessary subcomponents to do their respective reinitialization.
The steps should resemble (be a subset, but preserve the order) of steps that Configuration::initialize does.

That way you won't need any hooks to notify from lower level entities to high level entities about any adjustment done.